Add bindings for georep APIs

devyanikota opened the pull request:
Signed-off-by: Devyani Kota <devyanikota@gmail.com>

devyanikota committed:
Add bindings for georep APIs

Signed-off-by: Devyani Kota <devyanikota@gmail.com>

Madhu-1 commented on glusterapi/georep.py:
```diff
@@ -1,58 +1,160 @@
+
+
+def extract_volume_ID(remotehost, remoteport, remotevol):
+    if not remotehost:
```
you don't need this, as you are validating ```socket.inet_aton(remotehost)```

change ```GlusterApiInvalidInputs("Invalid IP Address")```
 to 

```
GlusterApiInvalidInputs("invalid remotehost address")
```

devyanikota replied:
done

Madhu-1 commented on glusterapi/georep.py:
```diff
@@ -1,58 +1,160 @@
+    try:
+        resp = requests.get(client_url)
+    except:
+        raise GlusterApiInvalidInputs("Invalid URL")
```
raise ```GlusterAPIError``` error

devyanikota replied:
Should this really be GlusterAPIError, considering that the client_url is formed from the user's input?
no?

Madhu-1 replied:
@aravindavk  whats your suggestion on this error?

Madhu-1 commented on glusterapi/georep.py:
```diff
@@ -1,58 +1,160 @@
         :param mastervol: (string) Master Volume Name
         :param remotehost: (string) Remote Host
+        :param remoteport: (string) Remote Port
         :param remotevol: (string) Remote Volume
```
Remote volume name

devyanikota replied:
done

Madhu-1 commented on glusterapi/georep.py:
```diff
@@ -1,58 +1,160 @@
+        mastervolid = resp.get(mastervol)
+        remotevolid = extract_volume_ID(mastervol, remotehost,
+                                        remoteport, remotevol)
+        url = "/geo-replication/%s/%s" % (mastervolid, remotevolid)
```
```url = "/v1/geo-replication/%s/%s" % (mastervolid, remotevolid)```

devyanikota replied:
rectified; thanks.

Madhu-1 commented on glusterapi/georep.py:
```diff
@@ -1,58 +1,160 @@
+        mastervolid = resp.get(mastervol)
+        remotevolid = extract_volume_ID(mastervol, remotehost, remoteport,
+                                        remotevol)
+        url = "/geo-replication/%s/%s/start" % (mastervolid, remotevolid)
```
url = "/v1/geo-replication/%s/%s" % (mastervolid, remotevolid)

Madhu-1 replied:
change URL in complete file. URL should start with /v1/

Madhu-1 commented on glusterapi/georep.py:
```diff
@@ -1,58 +1,160 @@
+    def georep_start(self, mastervol, remotehost, remoteport, remotevol,
+                     force=False):
         """
         Create Geo-replication Session
```
Start geo-replication Session

Madhu-1 commented on glusterapi/georep.py:
```diff
@@ -1,58 +1,160 @@
+        validate_volume_name(mastervol)
+        volumes_url = "/v1/volumes"
+        # Handle API response
+        try:
```
try block is not required here

Madhu-1 commented on glusterapi/georep.py:
```diff
@@ -1,58 +1,160 @@
+        validate_volume_name(mastervol)
+        volumes_url = "/v1/volumes"
+        # Handle API response
+        try:
```
try block is not required here

Madhu-1 commented on glusterapi/georep.py:
```diff
@@ -111,15 +213,29 @@ def georep_checkpoint(self, mastervol, remotehost,
+                resp = self._get(self, volumes_url)
+            except:
+                raise GlusterApiInvalidInputs("Invalid URL")
+            mastervolid = resp.get(mastervol)
```
try is required for ```mastervolid = resp.get(mastervol)```. not above

devyanikota replied:
done.

aravindavk commented on glusterapi/georep.py:
```diff
@@ -1,58 +1,149 @@
+        volumes_url = "/v1/volumes"
+        resp = self._get(self, volumes_url)
+        try:
+            mastervolid = resp.get(mastervol)
```
resp will have list of volinfo. `resp.get` always fails?

devyanikota replied:
@aravindavk 
the response received with URL: "/v1/volumes", is:
```
{"testvol1":"d2d8cb9f-5846-4556-94ee-1dddcdc54f44","testvol2":"dc81db34-a804-4727-a96a-de050faaa5db"}
```
while response received with URL: "/v1/volumes/testvol1", is:
"
{"ID":"d2d8cb9f-5846-4556-94ee-1dddcdc54f44","Name":"testvol1","Type":0,"Transport":"tcp","DistCount":1,"ReplicaCount":1,"Options":{},"Status":2,"Checksum":0,"Version":0,"Bricks":[{"Hostname":"10.70.42.62","NodeID":"946808b9-0296-44b6-82af-a2f7c5bc0305","Path":"/mnt/brick1","VolumeName":"testvol1","VolumeID":"d2d8cb9f-5846-4556-94ee-1dddcdc54f44"}],"Auth":{"Username":"0df46cb9-3cc1-4e7e-b70b-d9b901eb2596","Password":"9db2d4e8-2a2b-432c-981f-a9def5cf5328"}}
"

The response would fail if the master volume doesn't exist on the client node.
I had the idea that using first URL, one can extract the volume_ID of any volume by querying using the volume_name, thus, checking for the master volume.
But, like you suggested using the second URL, the extraction is limited to one specific volume; whose volume_ID can be extracted by querying for "ID" key.

The drawback was that one cannot query for remote vol id from this response on master node, correct?
So, will use georep status to extract both master and remote volume's ID ?
Correct me if am wrong.

aravindavk commented on glusterapi/georep.py:
```diff
@@ -1,58 +1,149 @@
+        :raises: GlusterAPIError or GlusterApiInvalidInputs on failure
+        """
+        validate_volume_name(mastervol)
+        volumes_url = "/v1/volumes"
```
get volinfo of specific volume using `/v1/volumes/<volname>`

aravindavk commented on glusterapi/georep.py:
```diff
@@ -1,58 +1,149 @@
-        :raises: GlusterAPIError or failure
-        """
-        pass
+    def georep_create(self, mastervol, remotehost, remoteport, remotevol,
```
Accept: `(mastervol, remotehost, remotevol, remoteuser="root", remote_endpoint="", remote_endpoint_user=None, remote_endpoint_secret=None, remote_endpoint_verify=False)`

First four arguments are for Geo-replication, other options are to connect to remote Cluster's Glusterd. (Similar to init function to connect to local Glusterd2 https://github.com/gluster/python-gluster-mgmt-client/blob/master/glusterapi/common.py#L33)

aravindavk commented on glusterapi/georep.py:
```diff
@@ -1,58 +1,149 @@
+            mastervolid = resp.get(mastervol)
+        except ValueError:
+            raise GlusterApiError("Master volume %s doesn't exist" % mastervol)
+        remotevolid = extract_volume_ID(mastervol, remotehost,
```
Send `remotevol, remote_endpoint, remote_endpoint_user, remote_endpoint_secret and remote_endpoint_verify`

aravindavk commented on glusterapi/georep.py:
```diff
@@ -1,58 +1,149 @@
+        :param remotevol: (string) Remote volume
+        :param force: (bool) Start Georep session with Force
+        :raises: GlusterAPIError or GlusterApiInvalidInputs on failure
+        """
```
- Get Geo-rep status.
- Extract Master Volume ID and Remote Volume ID from status output
- Call Geo-replication Start

No need to call `extract_volume_ID` since it makes call to remote Cluster. Once Session is established, these details are available with Geo-replication Status

aravindavk replied:
Getting all Status before every Geo-rep start/stop/delete may become expensive. We can introduce Geo-rep list API in future.(Like Volume info)

Madhu-1 replied:
as @aravindavk suggested we need ```remoteport``` and ```remotehost``` only for geo-replication create.
, you can make similar changes to other API's as well.

devyanikota replied:
done.

aravindavk commented on glusterapi/georep.py:
```diff
@@ -1,58 +1,149 @@
+        :param remotevol: (string) Remote volume
+        :param force: (bool) Stop Georep session with Force
+        :raises: GlusterAPIError or GlusterApiInvalidInputs on failure
+        """
```
Same as above. Get master volume id and remote volume id using status

devyanikota replied:
done

aravindavk commented on glusterapi/georep.py:
```diff
@@ -1,58 +1,149 @@
+        :param remotevol: (string) Remote volume
+        :param force: (bool) Delete Georep session with Force
+        :raises: GlusterAPIError or GlusterApiInvalidInputs on failure
+        """
```
Same as above. Get master volume id and remote volume id using status

devyanikota replied:
done.

aravindavk commented on glusterapi/georep.py:
```diff
@@ -111,15 +202,28 @@ def georep_checkpoint(self, mastervol, remotehost,
-        :raises: GlusterAPIError or failure
-        """
-        pass
+    def georep_status(self, mastervol=None, remotehost=None, remoteport=None,
```
Remove remoteport for now. and only implement else part now.

Once Geo-rep list/info API is implemented, then that can be used to get master volume id and Remote volume ID.

Madhu-1 replied:
This API is already there if we can list geo-rep sessions by```HTTP.GET``` ```/v1/geo-replication```

devyanikota replied:
done.

aravindavk reviewed (changes requested):
Sorry for the delayed review.

aravindavk commented:
Glusterd2 issue to get Geo-rep sessions list https://github.com/gluster/glusterd2/issues/950

devyanikota committed:
Add functionality for extraction of mastervolid and remotevolid

Use GeorepStatus API to enable extraction of mastervolid and
remotevolid.

Signed-off-by: Devyani Kota <devyanikota@gmail.com>

devyanikota committed:
Edit georep bindings to use GeorepStatus APIs to extract vol IDs

Signed-off-by: Devyani Kota <devyanikota@gmail.com>

devyanikota committed:
Add checks for authentication of valid clients

Signed-off-by: Devyani Kota <devyanikota@gmail.com>

devyanikota committed:
Add tests for geo-rep operations

Signed-off-by: Devyani Kota <devyanikota@gmail.com>

devyanikota commented:
@sac @aravindavk @Madhu-1 hi o/
I have added the georep tests as well.
Please check, thanks.

Madhu-1 commented on glusterapi/georep.py:
```diff
@@ -1,58 +1,149 @@
+    except socket.error:
+        raise GlusterApiInvalidInputs("Invalid Remote host IP Address")
+    validate_volume_name(remotevol)
+    client_url = "http://"+ remotehost + ":" + remoteport + "/v1/volumes"
```
we should not hardcode the schema, we may have to support both ```http``` and ```https```

Madhu-1 commented on glusterapi/georep.py:
```diff
@@ -1,58 +1,149 @@
+    except ValueError:
+        raise GlusterApiInvalidInputs("Invalid URL: %s" % client_url)
+    if resp.status_code == 200:
+        remotevolid = resp.json().get(remotevol)
```
what happens if  `remotevol` is not present

Madhu-1 commented on glusterapi/georep.py:
```diff
@@ -1,58 +1,149 @@
+    for s in allsessions:
+        if s.MasterID == mastervol:
+            mastervolid = s.MasterID
+    if mastervolid is None:
```
`mastervolid` is declared inside if statement, this may fail for the undeclared variable

Madhu-1 commented on glusterapi/georep.py:
```diff
@@ -1,58 +1,149 @@
+            for host in s.RemoteHosts:
+                if host.Hostname == remotehost:
+                    remotevolid = s.RemoteID
+    if remotevolid is None:
```
sam here for `remotevolid`

Madhu-1 commented on glusterapi/georep.py:
```diff
@@ -1,58 +1,149 @@
+        self.__init__(remote_endpoint, remote_endpoint_user,
+                      remote_endpoint_secret, remote_endpoint_verify)
+        headers = self._set_token_in_header('GET', volumes_url)
+        remotevolid = extract_volume_ID(remotehost, remoteport, remotevol, headers)
```
you should use `remote_endpoint` to get remote volume ID

Madhu-1 commented on glusterapi/georep.py:
```diff
@@ -1,58 +1,149 @@
+            mastervolid = resp.get(mastervol)
+        except ValueError:
+            raise GlusterApiError("Master volume %s doesn't exist" % mastervol)
+        self.__init__(remote_endpoint, remote_endpoint_user,
```
you are reinitializing Client class with remote_endpoint, are you able to create a geo-rep session from this?

Madhu-1 commented on glusterapi/georep.py:
```diff
@@ -1,58 +1,149 @@
+            "remotevolume": remotevol,
+            "remoteuser": remoteuser
+        }
+        return self._handle_request(self._post, httplib.CREATED,
```
there is a PR in gd2 to return HTTP status code `200` for some scenarios , need to check do you get `200` or `201` in this case

Madhu-1 commented on test/api_test.py:
```diff
@@ -79,3 +81,70 @@ def test_peer_remove(gd2client):
+    remotehost = remotehostinfo[0]
+    remoteport = remotehostinfo[1]
+    _, resp = gd2client.georep_create(mastervol, remotehost, remoteport, remotevol,
+                                      remoteuser="root", remote_endpoint="",
```
need to pass `remote_endpoint` as this should be used of geo rep creation.

devyanikota replied:
@Madhu-1 , I am passing the remote_endpoint!

Madhu-1 commented on test/api_test.py:
```diff
@@ -79,3 +81,70 @@ def test_peer_remove(gd2client):
+                                      remote_endpoint_user=gd2client.user,
+                                      remote_endpoint_secret=gd2client.secret,
+                                      remote_endpoint_verify=gd2client.verify)
+    assert bool(resp)
```
why are you doing assert with bool

Madhu-1 commented on test/api_test.py:
```diff
@@ -79,3 +81,70 @@ def test_peer_remove(gd2client):
+    # start geo-rep session
+    status, _ = gd2client.georep_start(mastervol, remotehost, remotevol,
+                                       force=False)
+    assert status == 200
```
as this assertion is already done at `georep_start` do we need to do it here?

Madhu-1 commented on test/api_test.py:
```diff
@@ -79,3 +81,70 @@ def test_peer_remove(gd2client):
+    _, resp = gd2client.peer_status()
+    peer_id = resp[0]["id"]
+    for i in range(3):
+        temp_path = tempfile.mkdtemp(prefix=str(i), dir="/tmp/")
```
why are you creating the directories in local machine?

devyanikota replied:
I wasn't aware of the fact that gd2 will be handling the creation of directories, in that case, what is expected to be sent against the bricks field? the default values will be used?

Madhu-1 commented on test/api_test.py:
```diff
@@ -79,3 +81,70 @@ def test_peer_remove(gd2client):
+
+    # delete the volume directories
+    for brick in m_bricks:
+        shutil.rmtree(brick)
```
we should not create or delete brick directories

devyanikota replied:
noted: will remove the creation and deletion operation of directories, since gd2 handles the same.

Madhu-1 reviewed (changes requested):
we need to come up with a better testing framework
* setup glusterd2 cluster
* do whatever we need ( volume create,volume list,volume delete etc)
* teardown glusterd2 cluster

devyanikota committed:
Handle switching to client session from remote session

Once the new georep connection is established between the
remote clusters, allow reinitializing of the client session
to interact with the client machine

Signed-off-by: Devyani Kota <devyanikota@gmail.com>

devyanikota commented:
@Madhu-1 I have handled the re-establishment of the client session after a successful georep connection is established. Removed the creation and deletion operations of directories(gd2 handles the same), and I have addressed the rest of the review comments on status checks.
Apart from this, added a TODO to track the task to organize tests using fixtures, which I can handle in the upcoming PR.
Please let me know if I need to take care of anything else.
@sac @Madhu-1 please review the same. Thank you.

devyanikota commented:
PR Status update: The tests added here for the georep APIs pass for the successful creation of georep sessions.
The tests for the rest of the georep operations when tested with the latest gd2 build are failing on the server end. I'll have to check on the gd2 end, for these raft errors. The gd2 binary version am using is the latest one, taken from the GD2 releases page[1].

```
[root@rhs31-b gd2_rpm]# ./glusterd2 --version
glusterd version: v4.1.0-0
```
[1] https://github.com/gluster/glusterd2/releases

If we can take this forward, I can work on the rest of the tests in the upcoming PR.
@sac @Madhu-1 what do you guys think?
Thanks.

devyanikota committed:
Allow input of georep values from the config json file

devyanikota committed:
Change protected function to public

Signed-off-by: Devyani Kota <devyanikota@gmail.com>

sac commented:
@devyanikota can you please abandon this PR and separate the API and the tests?
Once the code-autogeneration is stable, we will use the tests and can replace the API with autogenerated code.

devyanikota commented:
@sac yes, I'll go ahead and do that.

devyanikota commented:
Closing this PR, in favour of PR #19 and #20 
@sac ^^^
